### PR TITLE
Implementation of trace_call and trace_callMany

### PR DESCRIPTION
**Pull request by kjazgar**
## Changes:
Implementation of trace_call and trace_callMany methods in TraceRpcModule.

## Types of changes

What types of changes does your code introduce?

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [x] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [x] Yes
- [ ] No

### CODE REVIEW

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs [hunk 1]
```diff
@@ -118,6 +119,19 @@ private async Task<JsonRpcResponse> ExecuteRequestAsync(JsonRpcRequest rpcReques
         {
             ParameterInfo[] expectedParameters = method.Info.GetParameters();
             string[] providedParameters = request.Params ?? Array.Empty<string>();
+            if (methodName == "trace_callMany")
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:122]: we don't want to check methodName in the main JsonRpc interface

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs [hunk 2]
```diff
@@ -258,7 +272,17 @@ private async Task<JsonRpcResponse> ExecuteRequestAsync(JsonRpcRequest rpcReques
                     {
                         if (providedParameter.StartsWith('[') || providedParameter.StartsWith('{'))
                         {
-                            executionParam = _serializer.Deserialize(new JsonTextReader(new StringReader(providedParameter)), paramType);
+                            if (paramType == typeof(Tuple<TransactionForRpc, string[]>[]))
+                            {
+                                TraceCallManyParametersWrapper executionParamTemp =
+                                    _serializer.Deserialize<TraceCallManyParametersWrapper>(
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:278]: we should handle it by parameters converters

File: src/Nethermind/Nethermind.Blockchain/Processing/BlockProcessor.BlockValidationTransactionsExecutor.cs [hunk 3]
```diff
@@ -57,7 +60,15 @@ public TxReceipt[] ProcessTransactions(Block block, ProcessingOptions processing
...
-                _transactionProcessor.ProcessTransaction(block, currentTx, receiptsTracer, processingOptions, _stateProvider);
+                if ((processingOptions & ProcessingOptions.Trace) != ProcessingOptions.None)
+                {
+                    _callAndRestoreTransactionProcessor.ProcessTransaction(block, currentTx, receiptsTracer, processingOptions, _stateProvider);
+                }
+                else
+                {
+                    _executeTransactionProcessor.ProcessTransaction(block, currentTx, receiptsTracer, processingOptions, _stateProvider);
+                }
+
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Processing/BlockProcessor.BlockValidationTransactionsExecutor.cs:71]: Would prefer either BlockTraceTransactionsExecutor, we can have a base class which can take ITransactionProcessorAdapter

File: src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs [hunk 4]
```diff
@@ -37,10 +37,11 @@ public Tracer(IStateProvider stateProvider, IBlockchainProcessor blockProcessor,
             _processingOptions = processingOptions;
         }
 
-        public Keccak Trace(Block block, IBlockTracer blockTracer)
+        public Keccak Trace(Block block, IBlockTracer blockTracer, ProcessingOptions processingOptions = ProcessingOptions.Trace)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs:40]: seems redundant to ctor parameter

File: src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ public class Tracer : ITracer
     {
         private readonly IStateProvider _stateProvider;
         private readonly IBlockchainProcessor _blockProcessor;
-        private readonly ProcessingOptions _processingOptions;
+        public ProcessingOptions _processingOptions;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs:31]: ???

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/BuildUpTransactionProcessorAdapter.cs [hunk 6]
```diff
@@ -31,5 +31,8 @@ public BuildUpTransactionProcessorAdapter(ITransactionProcessor transactionProce
         
         public void Execute(Transaction transaction, BlockHeader block, ITxTracer txTracer) =>
             _transactionProcessor.BuildUp(transaction, block, txTracer);
+
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Evm/TransactionProcessing/BuildUpTransactionProcessorAdapter.cs:34]: missing the point, adapters where introduced to segregate the type of transaction execution

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcConfig.cs [hunk 7]
```diff
@@ -42,7 +42,7 @@ public int WebSocketsPort
         public string? IpcUnixDomainSocketPath { get; set; } = null;
 
         public string[] EnabledModules { get; set; } = ModuleType.DefaultModules.ToArray();
-        public long? GasCap { get; set; } = 100000000;
+        public long? GasCap { get; set; } = 50000000;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/JsonRpcConfig.cs:45]: Why are we dropping GasCap?

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs (hunk 1 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs:122]: code smell, if you need to change deserialization logic we need to do it by providing JsonConverter

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ITraceRpcModule.cs [hunk 8]
```diff
@@ -27,7 +28,7 @@ public interface ITraceRpcModule : IRpcModule
         ResultWrapper<ParityTxTraceFromReplay> trace_call(TransactionForRpc message, string[] traceTypes, BlockParameter blockParameter);
         
         [JsonRpcMethod(Description = "", IsImplemented = false, IsSharable = false)]
-        ResultWrapper<ParityTxTraceFromReplay[]> trace_callMany((TransactionForRpc message, string[] traceTypes, BlockParameter numberOrTag)[] calls);
+        ResultWrapper<ParityTxTraceFromReplay[]> trace_callMany(Tuple<TransactionForRpc, string[]>[] calls, BlockParameter numberOrTag);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ITraceRpcModule.cs:31]: Why Tuple and not class? Might save us issue with deserialization.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs [hunk 9]
```diff
@@ -63,15 +63,49 @@ private static ParityTraceTypes GetParityTypes(string[] types)
...
+            if (message.Gas == null || message.Gas == 0)
+            {
+
+                message.Gas = _jsonRpcConfig.GasCap ?? long.MaxValue;
+
+            }
+            else
+            {
+                message.Gas = Math.Min(_jsonRpcConfig.GasCap ?? long.MaxValue, message.Gas.Value);
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs:79]: This looks like some logic that is/should be reused in other places too?

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs [hunk 10]
```diff
@@ -63,15 +63,49 @@ private static ParityTraceTypes GetParityTypes(string[] types)
...
+                    txForRpc.Gas = _jsonRpcConfig.GasCap ?? long.MaxValue;
+                }
+                else
+                {
+                    txForRpc.Gas = Math.Min(_jsonRpcConfig.GasCap ?? long.MaxValue, txForRpc.Gas.Value);
+                }
+                Transaction tx = txForRpc.ToTransaction();
+                string[] traceTypes = obj.Item2;
+
+                ResultWrapper<ParityTxTraceFromReplay> trace = TraceTx(tx, traceTypes, numberOrTag);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs:104]: Can this just call trace_call?

**Review by LukaszRozmej — CHANGES_REQUESTED** [review]

File: src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs (hunk 1 above)
Comment (kjazgar) [reply src/Nethermind/Nethermind.JsonRpc/JsonRpcService.cs]: Added converter

File: src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs (hunk 4 above)
Comment (kjazgar) [reply src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs]: Deleted

File: src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs (hunk 5 above)
Comment (kjazgar) [reply src/Nethermind/Nethermind.Blockchain/Tracing/Tracer.cs]: fixed

File: src/Nethermind/Nethermind.Evm/TransactionProcessing/BuildUpTransactionProcessorAdapter.cs (hunk 6 above)
Comment (kjazgar) [reply src/Nethermind/Nethermind.Evm/TransactionProcessing/BuildUpTransactionProcessorAdapter.cs:34]: Removed unnecessary method.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ITraceRpcModule.cs (hunk 8 above)
Comment (kjazgar) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Trace/ITraceRpcModule.cs]: Changed.

File: src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs (hunk 10 above)
Comment (kjazgar) [reply src/Nethermind/Nethermind.JsonRpc/Modules/Trace/TraceRpcModule.cs]: Changed.

File: src/Nethermind/Nethermind.JsonRpc.Test/Modules/TraceRpcModuleTests.cs [hunk 11]
```diff
@@ -562,13 +527,137 @@ public async Task trace_replayTransaction_reward_test()
...
+            await context.Build();
+            string serialized = RpcTest.TestSerializedRequest(
+                EthModuleFactory.Converters.Union(TraceModuleFactory.Converters).ToList(), context.TraceRpcModule,
+                "trace_call", request1, request2);
+
+            Assert.AreEqual(expectedResult, serialized, serialized.Replace("\"", "\\\""));
+        }
+
+        [Test]
+        public async Task trace_callMany_test()
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.JsonRpc.Test/Modules/TraceRpcModuleTests.cs:610]: Please add more descriptive names to those tests

**Review by LukaszRozmej — APPROVED** [review]

**Review by MarekM25 — APPROVED** [review]

**Pull request merged**